src/test/java/org/opentripplanner/routing/algorithm/FilterTest.java

```diff
@@ -0,0 +1,485 @@
+package org.opentripplanner.routing.algorithm;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.opentripplanner.transit.model._data.TransitModelForTest.id;
+
+import java.util.List;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.request.RouteRequestTransitDataProviderFilter;
+import org.opentripplanner.routing.api.request.request.filter.SelectRequest;
+import org.opentripplanner.routing.api.request.request.filter.TransitFilterRequest;
+import org.opentripplanner.routing.core.RouteMatcher;
+import org.opentripplanner.transit.model._data.TransitModelForTest;
+import org.opentripplanner.transit.model.basic.MainAndSubMode;
+import org.opentripplanner.transit.model.basic.SubMode;
+import org.opentripplanner.transit.model.basic.TransitMode;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+
+public class FilterTest {
+
+  final String AGENCY_1 = "RUT:Agency:1";
+  final String AGENCY_2 = "RUT:Agency:2";
+  final String AGENCY_3 = "RUT:Agency:3";
+
+  final String ROUTE_1 = "RUT:Route:1";
+  final String ROUTE_2 = "RUT:Route:2";
+  final String ROUTE_3 = "RUT:Route:3";
+  final String ROUTE_4 = "RUT:Route:4";
```

Comment:
This is a very nice test, but it is not DRY, I would add constants for the 3 agencies used in almost every test case. The routes vary, so those I would create in the test as you did.

```suggestion
  final String AGENCY_ID_1 = "RUT:Agency:1";
  :
  final String ROUTE_ID_1 = "RUT:Route:1";
  :
  final Agency AGENCY_1 = TransitModelForTest.agency("A").copy().withId(id(AGENCY_ID_1)).build();
```
Note! Ids should also include id in the name - I know I probably did this mistake myself in other tests :-)

Reply:
The test does not have overlapping sets. A, B and C should have some of the same elements. Also I think Test 5 is wrong: 

`filters: [ { select: [ {A} ] }, { not: [ {B} ] } ] -> A ∪ (S - B)` should be

`filters: [ { select: [ {A} ] }, { not: [ {B} ] } ] -> A - B`. 

From spec: 
- The order of the TripFilterInput matter! I the select is not present in the TripFilterInput element, then we will remove the elements from what we got so fare. If we start with a "not" then we will take EVERYTHING minus the "not".

Reply:
`The order of the TripFilterInput matter! I the select is not present in the TripFilterInput element, then we will remove the elements from what we got so fare. If we start with a "not" then we will take EVERYTHING minus the "not".`

Not sure about that. That's 3 different behaviours depending on if `not` is combined with `select` or if it's first filter in the list or not. Does not feel intuitive for me.

Reply:
```
trip (filter : [
 { select : [ { feeds: ["DK"], modes: ["rail"] } ] },
 { not : [ { modes: ["water"] } ] }
 { select : [ { feeds: ["SE"] } ], not : [ { authorities: ["DR"] } ] },
])
```

For example, the case above. It's either feed DK with mode rail or feed SE with authority DR. Then we have this extra filter in the middle. According to the spec it should be applied together with first filter but not with the last. But that's only because we do not have any `select`. If I would add one then we would have different behaviour. Feels very chaotic to me.

Reply:
Agencies are not constants in the test. I was thinking about having common data set for all the tests but it would have to be quite big since we are many combinations we are testing. I the tests are easier to understand the way they are written right now.